### PR TITLE
chore: move python version compatibility to 3.10

### LINKED ISSUE #729: Bump minimum Python version to 3.10
### What needs to get done

Bump minimum Python version to 3.10

### Why it needs to get done

Because.

### PR DESCRIPTION
**Pull request by cmatsuoka**
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

### CODE REVIEW

**Comment by cmatsuoka** [comment]
Rebased on features/2.0

**Review by mr-cal — APPROVED** [review]
Looks good, a few non-blocking thoughts:

1. Can you use `from collections import namedtuple` instead of `from typing import NamedTuple`?
2. There's a leftover `Optional` here: https://github.com/canonical/craft-parts/blob/78bbc6a98dd9ea495487fd580ef7a7eefbaa274b/craft_parts/overlays/layers.py#L45
3. There's a leftover `Union` here: https://github.com/canonical/craft-parts/blob/78bbc6a98dd9ea495487fd580ef7a7eefbaa274b/craft_parts/packages/apt_cache.py#L395

**Comment by lengau** [comment]
Re: @mr-cal 's #1: I don't like the location of `typing.NamedTuple`, but it's better to use when we're using strong typing. Ideally we'd replace most or all of these with dataclasses in a separate PR.

File: craft_parts/plugins/rust_plugin.py [hunk 1]
```diff
@@ -37,7 +37,7 @@
 # see https://github.com/samuelcolvin/pydantic/issues/975#issuecomment-551147305
 # The proper fix requires Python 3.9+ (needs `typing.Annotated`)
 if TYPE_CHECKING:
```
Comment (lengau) [inline craft_parts/plugins/rust_plugin.py:39]: GitHub's not letting me apply the suggestion to the full thing, but:

```suggestion
UniqueStrList Annotated[liststr], conlist(str, unique_items=True)]
```

**Review by lengau — APPROVED** [review]
LGTM! The one thing I caught doesn't really need changing because it goes away with pydantic 2 anyway.

**Comment by cmatsuoka** [comment]
Removed remaining `Optional` and `Union` instances, typing adjustments to be handled separately.

**Pull request merged**